pkg/scheduler/algorithm/predicates/predicates_test.go

```diff
@@ -5039,3 +5040,445 @@ func TestCheckNodeUnschedulablePredicate(t *testing.T) {
+			// as the incoming pod doesn't have label "foo"
+			name: "pods spread across nodes as 2/1/0/3, but pod doesn't match itself",
+			pod: st.MakePod().Name("p").Label("bar", "").SpreadConstraint(
+				1, "node", hardSpread, st.MakeLabelSelector().Exists("foo").Obj(),
+			).Obj(),
+			nodes: []*v1.Node{
+				st.MakeNode().Name("node-a").Label("zone", "zone1").Label("node", "node-a").Obj(),
+				st.MakeNode().Name("node-b").Label("zone", "zone1").Label("node", "node-b").Obj(),
+				st.MakeNode().Name("node-x").Label("zone", "zone2").Label("node", "node-x").Obj(),
+				st.MakeNode().Name("node-y").Label("zone", "zone2").Label("node", "node-y").Obj(),
+			},
+			existingPods: []*v1.Pod{
+				st.MakePod().Name("p-a1").Node("node-a").Label("foo", "").Obj(),
+				st.MakePod().Name("p-a2").Node("node-a").Label("foo", "").Obj(),
+				st.MakePod().Name("p-b1").Node("node-b").Label("foo", "").Obj(),
+				st.MakePod().Name("p-y1").Node("node-y").Label("foo", "").Obj(),
+				st.MakePod().Name("p-y2").Node("node-y").Label("foo", "").Obj(),
+				st.MakePod().Name("p-y3").Node("node-y").Label("foo", "").Obj(),
+			},
+			fits: map[string]bool{
+				"node-a": false,
+				"node-b": true,
+				"node-x": true,
+				"node-y": false,
+			},
+		},
+		{
+			// only node-a and node-y are considered, so pods spread as 2/~1~/~0~/3
+			// ps: '~num~' is a markdown symbol to denote a crossline through 'num'
+			// but in this unit test, we don't run NodeAffinityPredicate, so node-b and node-x are
+			// still expected to be fits;
+			// the fact that node-a fits can prove the underlying logic works
+			name: "incoming pod has nodeAffinity, pods spread as 2/~1~/~0~/3, hence node-a fits",
+			pod: st.MakePod().Name("p").Label("foo", "").
+				NodeAffinityIn("node", []string{"node-a", "node-y"}).
+				SpreadConstraint(1, "node", hardSpread, st.MakeLabelSelector().Exists("foo").Obj()).
+				Obj(),
+			nodes: []*v1.Node{
+				st.MakeNode().Name("node-a").Label("zone", "zone1").Label("node", "node-a").Obj(),
+				st.MakeNode().Name("node-b").Label("zone", "zone1").Label("node", "node-b").Obj(),
+				st.MakeNode().Name("node-x").Label("zone", "zone2").Label("node", "node-x").Obj(),
+				st.MakeNode().Name("node-y").Label("zone", "zone2").Label("node", "node-y").Obj(),
+			},
+			existingPods: []*v1.Pod{
+				st.MakePod().Name("p-a1").Node("node-a").Label("foo", "").Obj(),
+				st.MakePod().Name("p-a2").Node("node-a").Label("foo", "").Obj(),
+				st.MakePod().Name("p-b1").Node("node-b").Label("foo", "").Obj(),
+				st.MakePod().Name("p-y1").Node("node-y").Label("foo", "").Obj(),
+				st.MakePod().Name("p-y2").Node("node-y").Label("foo", "").Obj(),
+				st.MakePod().Name("p-y3").Node("node-y").Label("foo", "").Obj(),
+			},
+			fits: map[string]bool{
+				"node-a": true,
+				"node-b": true, // in real case, it's false
+				"node-x": true, // in real case, it's false
+				"node-y": false,
+			},
+		},
+	}
+	for _, tt := range tests {
```

Comment:
can we add a test where an existing pod has labels that matches only one constraint of an incoming pod, basically test the case we discussed in https://github.com/kubernetes/kubernetes/pull/77760#discussion_r305609112

Reply:
They're added in the test `TestEvenPodsSpreadPredicate_MultipleConstraints`. You can check out commit "EvenPodsSpread: update 'selfMatch' logic".

Reply:
nice!